LibGfx: Prevent out of bounds access when scaling small Bitmaps

dariusarnold opened the pull request:
Fix #16047 
Since the color interpolation requires two pixels in the horizontal and vertical direction to work, 1 pixel wide or high bitmaps would cause a crash by out of bounds access when scaling. Fix this by clamping the index into the valid range.

I did not create test cases for the <1 case, because I could not create a Bitmap with either dimension as 0. The allocator failed to allocate the 0 byte storage for it. If I overlooked something and there is a way to create Bitmaps with any dimension as 0, I can add tests for it.

Overriding the get_pixel function could be done conditionally (only override if any dimension is 1) to improve performance, but I was not sure about the tradeoff between simplicity and speed. If avoiding all the comparisons in clamp() is worth the slight increase in complexity, I would be happy to change the PR.

Zaggy1024 commented:
Doesn't it make more sense for these to be fast paths instead?
- If Bitmap is 1x1, just copy the image
- If Bitmap is x pixels wide, 1 pixel tall, skip all but the bottom strip portion
- If Bitmap is x pixels tall, 1 pixel wide, skip all but the right strip portion

dariusarnold commented:
I hope I understood your suggestion correctly. I now implemented multiple paths, depending on the dimension of the original bitmap.

Zaggy1024 commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,111 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
 
-        int i = floorf(p);
-        float u = p - static_cast<float>(i);
+    } else if (old_height == 1 && old_width >= 1) {
```
Nits: If you verify that the bitmap doesn't have any 0-length dimensions (which I assume shouldn't be possible), you're guaranteed here that (old_height > 1 || old_width > 1), so you can use early returns to guarantee you only need to check one condition here:

```
if (old_width > 1 && old_height > 1) {
    ...
    return;
}
if (old_height == 1) {
    // Horizontal strips
    return;
}
// Vertical strips
```

dariusarnold replied:
Good suggestion, thank you.

Zaggy1024 commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,111 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
+        // Copy horizontal strip multiple times (excluding last pixel)
+        auto old_bottom_y = old_height - 1;
+        for (int new_bottom_y = 0; new_bottom_y < new_height; new_bottom_y++) {
+            for (int x = 0; x < new_width - 1; x++) {
```
You should be able to swap these loops to interpolate the pixels only once per row/column, and set the final row/column to the last pixel in another loop.

dariusarnold replied:
I couldn't figure out how swapping the loops alone would help, but since interpolation is only required to be done once for a row I now copy the interpolated value into all columns at once. I think that was what you meant with your comment.
```
        for (int x = 0; x < new_width - 1; x++) {
            auto p = static_cast<float>(x) * static_cast<float>(old_width - 1) / static_cast<float>(new_width - 1);
            int i = floorf(p);
            float u = p - static_cast<float>(i);

            auto a = get_pixel(i, old_bottom_y);
            auto b = get_pixel(i + 1, old_bottom_y);
            auto color = a.interpolate(b, u);
            for (int new_bottom_y = 0; new_bottom_y < new_height; new_bottom_y++) {
                // Interpolate color only once and then copy into all columns.
                new_bitmap->set_pixel(x, new_bottom_y, color);
            }
        }
```

Zaggy1024 replied:
That's exactly what I meant, thank you! sorry if my explanation was considering.

Zaggy1024 commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,111 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
+            // Copy last pixel of horizontal strip
+            new_bitmap->set_pixel(new_width - 1, new_bottom_y, get_pixel(physical_width() - 1, old_bottom_y));
+        }
+    } else if (old_height > 1 && old_width == 1) {
```
Nits: See if statement comment above.

Zaggy1024 commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,111 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
-    auto new_right_x = new_width - 1;
-    for (int y = 0; y < new_height - 1; y++) {
-        auto q = static_cast<float>(y) * static_cast<float>(old_height - 1) / static_cast<float>(new_height - 1);
+            for (int y = 0; y < new_height - 1; y++) {
```
Same for this set of loops.

Zaggy1024 reviewed (changes requested):
Just a couple more suggestions, and then this should be good in my opinion.

dariusarnold commented:
I also fixed the failing 32bit unit tests. It seems that the float rounding is different between architectures that scaling by the same value produces different results. I adjusted the scaling factors to not land on that edge.

Zaggy1024 reviewed (approved):
LGTM! I'm sure maintainers will need a look too.

not-a-skeleton commented:
Skipping the redundant interpolations in x-by-1 and 1-by-y cases does seem like a plausible optimization, but the special cases add a fair bit of complexity, IMO. Have we measured them to confirm this extra code is actually profitable?

FWIW here's an alternate patch that would fix the crash without adding any branches in the for loops.

```diff
--- a/Userland/Libraries/LibGfx/Bitmap.cpp
+++ b/Userland/Libraries/LibGfx/Bitmap.cpp
@@ -389,6 +389,9 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
     auto new_width = new_bitmap->physical_width();
     auto new_height = new_bitmap->physical_height();

+    int adjacent_x_offset = old_width > 1 ? 1 : 0;
+    int adjacent_y_offset = old_height > 1 ? 1 : 0;
+
     // The interpolation goes out of bounds on the bottom- and right-most edges.
     // We handle those in two specialized loops not only to make them faster, but
     // also to avoid four branch checks for every pixel.
@@ -404,9 +407,9 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
             float v = q - static_cast<float>(j);

             auto a = get_pixel(i, j);
-            auto b = get_pixel(i + 1, j);
-            auto c = get_pixel(i, j + 1);
-            auto d = get_pixel(i + 1, j + 1);
+            auto b = get_pixel(i + adjacent_x_offset, j);
+            auto c = get_pixel(i, j + adjacent_y_offset);
+            auto d = get_pixel(i + adjacent_x_offset, j + adjacent_y_offset);

             auto e = a.interpolate(b, u);
             auto f = c.interpolate(d, u);
@@ -425,7 +428,7 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
         float u = p - static_cast<float>(i);

         auto a = get_pixel(i, old_bottom_y);
-        auto b = get_pixel(i + 1, old_bottom_y);
+        auto b = get_pixel(i + adjacent_x_offset, old_bottom_y);
         auto color = a.interpolate(b, u);
         new_bitmap->set_pixel(x, new_bottom_y, color);
     }
@@ -440,7 +443,7 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
         float v = q - static_cast<float>(j);

         auto c = get_pixel(old_right_x, j);
-        auto d = get_pixel(old_right_x, j + 1);
+        auto d = get_pixel(old_right_x, j + adjacent_y_offset);

         auto color = c.interpolate(d, v);
         new_bitmap->set_pixel(new_right_x, y, color);
```

not-a-skeleton commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,114 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
+        // Bottom-right pixel
+        new_bitmap->set_pixel(new_width - 1, new_height - 1, get_pixel(physical_width() - 1, physical_height() - 1));
+        return new_bitmap;
+    } else if (old_height == 1) {
```
Nit: since we `return` on the line above, this could be an `if` rather than an `else if`.

not-a-skeleton commented on Userland/Libraries/LibGfx/Bitmap.cpp:
```diff
@@ -389,66 +389,114 @@ ErrorOr<NonnullRefPtr<Gfx::Bitmap>> Bitmap::scaled(float sx, float sy) const
+            new_bitmap->set_pixel(new_width - 1, new_bottom_y, get_pixel(physical_width() - 1, old_bottom_y));
+        }
+        return new_bitmap;
+    } else if (old_width == 1) {
```
Nit: since we `return` on the line above, this could be an `if` rather than an `else if`.
